shared/rakefiles/xk.rake

```diff
@@ -0,0 +1,175 @@
+
+@gcp_creds_file = "/root/.config/gcloud/credentials.db"
+task :configure_login => [@gcp_creds_file]
+rule @gcp_creds_file do
+  # This authenticates to GCP/the `gcloud` command in the normal, interactive
+  # way. This is the default method, best for human users.
+  sh "gcloud auth login"
+end
+
+# A more rake-ish way to create this file would be:
+#
+#   task :configure_serviceaccount => [@serviceaccount_key_file]
+#   rule @serviceaccount_key_file => [@gcp_creds_file] do ... end
+#
+# However, that logic would force creation of a new @serviceaccount_key_file
+# (and a new Serviceaccount Key) whenever @gcp_creds_file changes. This is not
+# what we want, i.e. don't create a new Key/key file when @gcp_creds_file
+# changes because of :configure_kubectl.
+@serviceaccount_key_file = ENV["TF_VAR_serviceaccount_key"]
+task :configure_serviceaccount => [@gcp_creds_file, :configure_current_project] do
+  # TODO: This command is duplicated from exekube's gcp-project-init (and
+  # hardcodes 'projectowner' instead of $SA_NAME which is only defined in
+  # gcp-project-init). If gcp-project-init becomes idempotent (GPII-2989,
+  # https://github.com/exekube/exekube/issues/92), or if this 'keys create'
+  # step moves somewhere else in exekube, call this command from that place
+  # instead.
+  sh "
+    [ -f $TF_VAR_serviceaccount_key ] || \
+    gcloud iam service-accounts keys create $TF_VAR_serviceaccount_key \
+      --iam-account projectowner@$TF_VAR_project_id.iam.gserviceaccount.com
+  "
+end
+
+@kubectl_creds_file = "/root/.kube/config"
+task :configure_kubectl => [@kubectl_creds_file]
+rule @kubectl_creds_file => [@gcp_creds_file] do
+  # This duplicates information in terraform code, 'k8s-cluster'
+  cluster_name = 'k8s-cluster'
+  # This duplicates information in terraform code, 'zone'. Could be a variable with some plumbing.
+  zone = 'us-central1-a'
+  sh "
+    existing_cluster=$(gcloud container clusters list --filter #{cluster_name} --zone #{zone} --project #{ENV["TF_VAR_project_id"]})
+    if [ $? == 0 ] && [ \"$existing_cluster\" != \"\" ]; then \
+      gcloud container clusters get-credentials #{cluster_name} --zone #{zone} --project #{ENV["TF_VAR_project_id"]}
+    fi"
+end
+
+task :configure_current_project => [@gcp_creds_file] do
+  sh "gcloud config set project $TF_VAR_project_id"
+end
+
+# This task is being called from entrypoint.rake and runs inside exekube container.
+# It applies secret-mgmt, sets secrets, and then executes arbitrary command from args[:cmd].
+# You should not invoke this task directly!
+task :xk, [:cmd, :skip_secret_mgmt] => [@serviceaccount_key_file, @kubectl_creds_file] do |taskname, args|
+  @secrets = Secrets.collect_secrets()
+
+  sh "#{@exekube_cmd} up live/#{@env}/secret-mgmt" unless args[:skip_secret_mgmt]
+
+  Secrets.set_secrets(@secrets)
+
+  sh_filter "#{@exekube_cmd} #{args[:cmd]}" if args[:cmd]
+end
+
+task :refresh_common_infra, [:project_type] => [@gcp_creds_file] do | taskname, args|
```

Comment:
I don't think that `xk.rake` is a right place for `:refresh_common_infra` and `:apply_common_infra` tasks. It is intended to be a lightweight wrapper for commands that require secrets to be set prior to execution. Is there anything that prevents moving those 2 tasks to common entrypoint rake file?

Reply:
Both tasks don't require secrets, they only need to be authenticated in GCP, that's why the `@gcp_creds_file` is a dependency.

Also they are invoked from [entrypoint.rake](https://github.com/gpii-ops/gpii-infra/blob/d2a4efb99a5dd9c02806d3ac16f92ff7f43244be/shared/rakefiles/entrypoint.rake#L79) and [entrypoint_common.rake](https://github.com/gpii-ops/gpii-infra/blob/d2a4efb99a5dd9c02806d3ac16f92ff7f43244be/shared/rakefiles/entrypoint-common.rake#L10) by tasks that set the secrets.

But if secrets were needed, I'd vote for creating a new task inside xk.rake in order to manage them and then use it as a dependency by other tasks. I think that one goal is to have as much as possible inside the Docker container.

I can not see the advantages of adding an extra layer to the sh commands. I mean using `rake -> terraform -> sh` instead of `rake -> sh`.

Reply:
@amatas Ok, I am still convinced that those tasks belong to entrypoint and entrypoint_common. However, I understand if you don't want to deal with this now and focus on merging this PR instead. We can fix this later.